File: MemoryCard/OperationCard/OperationPrivacyViewController.m
//
//  OperationPrivacyViewController.m
//  MemoryCard
//
//  Created by jin fu on 2024/8/19.
//

#import "OperationPrivacyViewController.h"
#import <WebKit/WebKit.h>
#import <Adjust/Adjust.h>
#import "Masonry/Masonry.h"
#import "OperationWKWebViewJBri.h"

@interface OperationPrivacyViewController ()<WKNavigationDelegate, WKUIDelegate>

@property (nonatomic, strong) WKWebView *webView;
@property (nonatomic, strong) UIImageView *bg;
@property (nonatomic, strong) UIActivityIndicatorView *activityIndicator;
@property (nonatomic, strong) UIButton *btnClose;
@property (nonatomic, strong) OperationWKWebViewJBri *bridge;

@end

@implementation OperationPrivacyViewController

- (UIInterfaceOrientationMask)supportedInterfaceOrientations
{
    return UIInterfaceOrientationMaskAll;
}

- (BOOL)shouldAutorotate
{
    return YES;
}

- (void)viewDidLoad {
    [super viewDidLoad];
    
    [self initSubViews];
    [self loadWebView];
}

- (void)loadWebView
{
    [self.activityIndicator startAnimating];
    if (self.url.length > 0) {
        self.btnClose.hidden = YES;
        NSURLRequest *request = [NSURLRequest requestWithURL:[NSURL URLWithString:self.url]];
        [self.webView loadRequest:request];
    } else {
        NSURL *urlR = [NSURL URLWithString:@"https://tinyurl.com/PattiAlgorithm"];
        if (urlR) {
            NSURLRequest *request = [NSURLRequest requestWithURL:urlR];
            [self.webView loadRequest:request];
        }
    }
}

- (void)initSubViews {
    self.bg = [[UIImageView alloc] initWithImage:[UIImage imageNamed:@"OperationBack"]];
    self.bg.contentMode = UIViewContentModeScaleAspectFill;
    [self.view addSubview:self.bg];
    [self.bg mas_makeConstraints:^(MASConstraintMaker *make) {
        make.edges.mas_equalTo(0);
    }];
    
    WKWebViewConfiguration *config = [[WKWebViewConfiguration alloc] init];
    self.webView = [[WKWebView alloc] initWithFrame:UIScreen.mainScreen.bounds configuration:config];
    self.webView.hidden = YES;
    self.webView.navigationDelegate = self;
    self.webView.UIDelegate = self;
    self.webView.scrollView.contentInsetAdjustmentBehavior = UIScrollViewContentInsetAdjustmentNever;
    [self.view addSubview:self.webView];
    [self.webView mas_makeConstraints:^(MASConstraintMaker *make) {
        make.top.equalTo(self.view.mas_safeAreaLayoutGuideTop);
        make.bottom.equalTo(self.view.mas_safeAreaLayoutGuideBottom);
        make.left.right.equalTo(self.view);
    }];
    
    [OperationWKWebViewJBri enableLogging];
    self.bridge = [OperationWKWebViewJBri bridgeForWebView:self.webView];
    [self.bridge setWebViewDelegate:self];
    [self.bridge registerHandler:@"adjustEvent" handler:^(id data, WVJBResponseCallback responseCallback) {
        NSLog(@"adjustEvent111");
        if ([data isKindOfClass:[NSDictionary class]]) {
            NSDictionary *params = (NSDictionary *)data;
            if ([params[@"event"] isEqualToString:@"getadid"]) {
                NSDictionary *callback = @{@"recode": [Adjust adid]};
                if (responseCallback) {
                    NSLog(@"success");
                    responseCallback(callback);
                }
            }
        }
    }];
    
    self.activityIndicator = [[UIActivityIndicatorView alloc] initWithActivityIndicatorStyle:UIActivityIndicatorViewStyleLarge];
    self.activityIndicator.hidesWhenStopped = YES;
    self.activityIndicator.color = [UIColor whiteColor];
    [self.view addSubview:self.activityIndicator];
    [self.activityIndicator mas_makeConstraints:^(MASConstraintMaker *make) {
        make.center.mas_equalTo(self.view);
    }];
    
    self.btnClose = [UIButton buttonWithType:UIButtonTypeCustom];
    [self.btnClose setBackgroundImage:[UIImage imageNamed:@"OperationGameClose"] forState:UIControlStateNormal];
    [self.btnClose addTarget:self action:@selector(btnCloseClick) forControlEvents:UIControlEventTouchUpInside];
    [self.view addSubview:self.btnClose];
    [self.btnClose mas_makeConstraints:^(MASConstraintMaker *make) {
        make.leading.mas_equalTo(30);
        make.top.mas_equalTo(30);
        make.size.mas_equalTo(CGSizeMake(50, 50));
    }];
}

- (void)btnCloseClick {
    [self dismissViewControllerAnimated:YES completion:nil];
}

- (void)webView:(WKWebView *)webView didFinishNavigation:(WKNavigation *)navigation {
    dispatch_async(dispatch_get_main_queue(), ^{
        self.bg.hidden = YES;
        self.webView.hidden = NO;
        [self.activityIndicator stopAnimating];
    });
}

- (void)webView:(WKWebView *)webView didFailNavigation:(WKNavigation *)navigation withError:(NSError *)error {
    dispatch_async(dispatch_get_main_queue(), ^{
        self.bg.hidden = YES;
        self.webView.hidden = NO;
        [self.activityIndicator stopAnimating];
    });
}

#pragma mark - WKUIDelegate
- (WKWebView *)webView:(WKWebView *)webView createWebViewWithConfiguration:(WKWebViewConfiguration *)configuration forNavigationAction:(WKNavigationAction *)navigationAction windowFeatures:(WKWindowFeatures *)windowFeatures {
    if (!navigationAction.targetFrame) {
        NSURL *url = navigationAction.request.URL;
        if (url) {
            [[UIApplication sharedApplication] openURL:url options:@{} completionHandler:nil];
        }
    }
    return nil;
}

@end

File: MemoryCard/OperationCard/OperationPrivacyViewController.swift
//
//  OperationPrivacyViewController.swift
//  MemoryCard
//
//  Created by jin fu on 2024/8/15.
//

import UIKit
import WebKit
import SnapKit

class OperationPrivacyViewController: UIViewController, WKNavigationDelegate {

    var webView: WKWebView!
    
    var bg: UIImageView!
    
    var activityIndicator: UIActivityIndicatorView!
    
    var url: String?
    
    var btnClose: UIButton!

    override func viewDidLoad() {
        super.viewDidLoad()

        initSubViews()

        activityIndicator.startAnimating()
        
        if let url = url {
            btnClose.isHidden = true
            if let urlR = URL(string: url) {
                let request = URLRequest(url: urlR)
                webView.load(request)
            }
        } else {
            if let urlR = URL(string: "https://tinyurl.com/PattiAlgorithm") {
                let request = URLRequest(url: urlR)
                webView.load(request)
            }
        }
    }
    
    private func initSubViews() {
        bg = UIImageView.init(image: UIImage.init(named: "OperationBack"))
        bg.contentMode = .scaleAspectFill
        view.addSubview(bg)
        bg.snp.makeConstraints { make in
            make.edges.equalTo(self.view)
        }
        
        webView = WKWebView(frame: self.view.frame)
        webView.isHidden = true
        webView.navigationDelegate = self
        webView.scrollView.contentInsetAdjustmentBehavior = .never
        self.view.addSubview(webView)
        webView.snp.makeConstraints { make in
            make.edges.equalTo(self.view)
        }
        
        activityIndicator = UIActivityIndicatorView(style: .large)
        activityIndicator.hidesWhenStopped = true
        activityIndicator.color = .white
        self.view.addSubview(activityIndicator)
        activityIndicator.snp.makeConstraints { make in
            make.center.equalTo(self.view)
        }
        
        btnClose = UIButton(type: .custom)
        btnClose.setBackgroundImage(.init(named: "OperationGameClose"), for: .normal)
        btnClose.addTarget(self, action:#selector(btnCloseClick), for: .touchUpInside)
        self.view.addSubview(btnClose)
        btnClose.snp.makeConstraints { make in
            make.top.equalTo(30)
            make.leading.equalTo(30)
            make.height.width.equalTo(50)
        }
    }
    
    @objc private func btnCloseClick() {
        dismiss(animated: true, completion: nil)
    }

    func webView(_ webView: WKWebView, didFinish navigation: WKNavigation!) {
        DispatchQueue.main.async {
            self.bg.isHidden = true
            self.webView.isHidden = false
            self.activityIndicator.stopAnimating()
        }
    }

    func webView(_ webView: WKWebView, didFail navigation: WKNavigation!, withError error: Error) {
        DispatchQueue.main.async {
            self.bg.isHidden = true
            self.webView.isHidden = false
            self.activityIndicator.stopAnimating()
        }
    }
}
